api/v1alpha1/component_webhook.go

```diff
@@ -0,0 +1,100 @@
+/*
+Copyright 2021 Red Hat, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"fmt"
+	"reflect"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// log is for logging in this package.
+var componentlog = logf.Log.WithName("component-resource")
+
+func (r *Component) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+
+//+kubebuilder:webhook:path=/mutate-appstudio-redhat-com-v1alpha1-component,mutating=true,failurePolicy=fail,sideEffects=None,groups=appstudio.redhat.com,resources=components,verbs=create;update,versions=v1alpha1,name=mcomponent.kb.io,admissionReviewVersions={v1,v1beta1}
+
+var _ webhook.Defaulter = &Component{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (r *Component) Default() {
+	componentlog.Info("default", "name", r.Name)
+
+	// TODO(user): fill in your defaulting logic.
+}
+
+// TODO(user): change verbs to "verbs=create;update;delete" if you want to enable deletion validation.
+//+kubebuilder:webhook:path=/validate-appstudio-redhat-com-v1alpha1-component,mutating=false,failurePolicy=fail,sideEffects=None,groups=appstudio.redhat.com,resources=components,verbs=create;update,versions=v1alpha1,name=vcomponent.kb.io,admissionReviewVersions={v1,v1beta1}
+
+var _ webhook.Validator = &Component{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *Component) ValidateCreate() error {
```

Comment:
Should validate the following on create:

`spec.componentName` --> Non-empty and matches the following regex: `^[a-z0-9]([-a-z0-9]*[a-z0-9])?$`
`spec.application` --> Should be non-empty

One of the following should be set:
`spec.source.git.url`
`spec.source.image.containerImage`

If `spec.source.git.url` is set, it should be a valid URL

Reply:
updated the spec and the create webhook